packer/provisioner.go

```diff
@@ -145,15 +145,61 @@ func (p *PausedProvisioner) Prepare(raws ...interface{}) error {
 
 func (p *PausedProvisioner) Provision(ctx context.Context, ui packersdk.Ui, comm packersdk.Communicator, generatedData map[string]interface{}) error {
 
-	// Use a select to determine if we get cancelled during the wait
 	ui.Say(fmt.Sprintf("Pausing %s before the next provisioner...", p.PauseBefore))
+
+	var err error
+	switch ui.(type) {
+	case *MachineReadableUi:
+		err = p.pausingNoUpdate(ctx)
+	default:
+		minimumTimeForPauseUpdate := float64(10)
+		if p.PauseBefore.Seconds() < minimumTimeForPauseUpdate {
+			err = p.pausingNoUpdate(ctx)
+		} else {
+			err = p.pausingWithUpdates(ctx, ui)
+		}
+	}
+	if err != nil {
+		return err
+	}
+
+	return p.Provisioner.Provision(ctx, ui, comm, generatedData)
+}
+
+func (p *PausedProvisioner) pausingNoUpdate(ctx context.Context) error {
 	select {
 	case <-time.After(p.PauseBefore):
 	case <-ctx.Done():
 		return ctx.Err()
 	}
+	return nil
+}
 
-	return p.Provisioner.Provision(ctx, ui, comm, generatedData)
+func (p *PausedProvisioner) pausingWithUpdates(ctx context.Context, ui packersdk.Ui) error {
+	updateTime := 10
+	timeTicker := time.NewTicker(time.Duration(updateTime) * time.Second)
+	TotalTime := p.PauseBefore.Seconds()
+	var err error
+	go func() {
+		for {
+			select {
+			case <-timeTicker.C:
+				TotalTime -= float64(updateTime)
+				if TotalTime != 0 {
+					ui.Say(fmt.Sprintf("%v seconds left until the next provisioner", TotalTime))
+				}
+				if TotalTime < float64(updateTime) {
+					return
+				}
+			case <-ctx.Done():
+				err = ctx.Err()
+				return
+			}
+		}
+	}()
+	time.Sleep(p.PauseBefore)
```

Comment:
I think it would be better if we used contexts and timeouts here as I think that the time pkg makes things just right and easy to read, here's an example that should just work :

https://play.golang.org/p/31HJA8xZodS

Reply:
Adjusted this function, let me know if there is anything else I should change.